lib/blocktree/blocktree.go

```diff
@@ -504,15 +526,54 @@ func (bt *BlockTree) DeepCopy() *BlockTree {
 }
 
 // StoreRuntime stores the runtime for corresponding block hash.
-func (bt *BlockTree) StoreRuntime(hash common.Hash, in Runtime) {
-	bt.runtimes.set(hash, in)
+func (bt *BlockTree) StoreRuntime(hash common.Hash, instance Runtime) {
+	bt.runtimes.set(hash, instance)
 }
 
-// GetBlockRuntime returns block runtime for corresponding block hash.
+// GetBlockRuntime returns block runtime for corresponding block hash, if there is no instance for
+// the given block hash it will lookup an instance of an ancestor and returns it.
 func (bt *BlockTree) GetBlockRuntime(hash common.Hash) (Runtime, error) {
-	ins := bt.runtimes.get(hash)
-	if ins == nil {
-		return nil, fmt.Errorf("%w for hash %s", ErrFailedToGetRuntime, hash)
+	// if the current node contains a runtime entry in the runtime mapping
+	// then we early return the instance, otherwise we will lookup for the
+	// closest parent with a runtime instance entry in the mapping
+	runtimeInstance := bt.runtimes.get(hash)
+	if runtimeInstance != nil {
+		return runtimeInstance, nil
+	}
+
+	bt.RLock()
+	defer bt.RUnlock()
+
+	currentNode := bt.getNode(hash)
+	if currentNode == nil {
+		return nil, fmt.Errorf("%w: %s", ErrNodeNotFound, hash)
+	}
+
+	currentNode = currentNode.parent
+	for currentNode != nil {
+		runtimeInstance := bt.runtimes.get(currentNode.hash)
+		if runtimeInstance != nil {
+			return runtimeInstance, nil
+		}
+
+		currentNode = currentNode.parent
```

Comment:
Why are we walking the whole chain here and then do something different in `blockState.closestAncestorWithInstance`?

Reply:
Good catch! I just removed this method, I don't remember why I put it there